### PR TITLE
State exercise

### PR DESCRIPTION
**Pull request by puyo**

### CODE REVIEW

File: src/Course/State.hs [hunk 1]
```diff
@@ -13,6 +13,8 @@ import Course.Functor
 import Course.Applicative
 import Course.Monad
 import qualified Data.Set as S
+import Debug.Trace
+import Data.Char
```
Comment (JulianLeviston) [inline src/Course/State.hs:17]: FYI, I think the functions we need are already imported. (`digitToInt`)

File: src/Course/State.hs [hunk 2]
```diff
@@ -26,61 +28,42 @@ import qualified Data.Set as S
...
 -- (10,6)
 instance Functor (State s) where
-  (<$>) ::
-    (a -> b)
-    -> State s a
-    -> State s b
-  (<$>) =
-    error "todo: Course.State#(<$>)"
+  (<$>) :: (a -> b) -> State s a -> State s b
+  (<$>) f state = State (\x -> (f $ eval state x, exec state x))
```
Comment (JulianLeviston) [inline src/Course/State.hs:66]: I did this by:

```
(<$>) f state = 
  State $ 
      \s -> 
          let (x, y) = runState state s
          in (f x, y)
```

Not sure if that is clearer, but it saves running the state twice, possibly (ie eval and exec run the state).

File: src/Course/State.hs [hunk 3]
```diff
@@ -94,17 +77,13 @@ instance Functor (State s) where
...
-    -> State s b 
-  (<*>) =
-    error "todo: Course.State (<*>)#instance (State s)"
+  pure :: a -> State s a
+  pure x = State (\y -> (x, y))
+  (<*>) :: State s (a -> b) -> State s a -> State s b
+  (<*>) s1 s2 = State (\x -> (e1 x, e2 x))
+    where
+      e1 x = eval s1 x (eval s2 x)
+      e2 x = exec s2 (exec s1 x)
```
Comment (JulianLeviston) [inline src/Course/State.hs:86]: So... the `s1` state has the function inside it... and the body of `e1` is evaluating s1, then applying that to `x` and the evaluation of s2 with x.

It's a bit tricky to read.... here's an alternative which you might find easier to understand:

```
    State (\s -> -- still needs to be inside a function on some data.. "s" here
      let
        (f, s2) = runState stateF s -- run the state to get the f out of stateF with the data value "s"
        (x, s3) = runState stateX s2 -- weave that last resultant data value "s2" back in to runState to get the "x" value out, that we'll apply "f" to in a second...
      in
        (f x, s3) -- the tuple we reply with has the "f" extracted from the first state applied to the "x" extracted from the second state, and the resultant value is the woven state that we got back from running state to get "x" out (ie the cumulative effected result value).
    )
```

WDYT?

File: src/Course/State.hs [hunk 4]
```diff
@@ -113,13 +92,21 @@ instance Applicative (State s) where
...
+
 instance Monad (State s) where
-  (=<<) ::
-    (a -> State s b)
-    -> State s a
-    -> State s b
-  (=<<) =
-    error "todo: Course.State (=<<)#instance (State s)"
+  (=<<) :: (a -> State s b) -> State s a -> State s b
+  (=<<) f state1 = State $ \s -> let (val1, s1) = runState state1 s in runState (f val1) s1 -- just copied this from the Internet
```
Comment (JulianLeviston) [inline src/Course/State.hs:109]: One liners are a bit tricky to understand... but sorry, that's obvious to you... you're just trying to get it work, as you say.

My version might be worth looking at, might help you understand it a bit more...

https://github.com/JulianLeviston/fp-course/pull/7/files#diff-ad52b034ecdacc8f3d5b1e973ac02c79R136

I think the key point is that in monad, we're taking each result and doing the next thing... sequencing, kinda.

File: src/Course/State.hs (hunk 3 above)
Comment (JulianLeviston) [reply src/Course/State.hs:86]: Also worth noting, we don't have to do the `(f, s2)` first, we could choose the `x` to do with `s2` first if we wanted.

File: src/Course/State.hs [hunk 5]
```diff
@@ -135,44 +122,137 @@ instance Monad (State s) where
...
-  error "todo: Course.State#findM"
+--
+--
+findM :: Monad f => (a -> f Bool) -> List a -> f (Optional a)
+findM _ Nil = return Empty
+findM p (x :. xs) =
+  p x
+  >>= \b -> if b then return (Full x) else findM p xs
+
+-- attempt 2: foldRight, but maybe this forces it to inspect every value in the list rather than short circuit?
```
Comment (JulianLeviston) [inline src/Course/State.hs:133]: 👍

File: src/Course/State.hs [hunk 6]
```diff
@@ -135,44 +122,137 @@ instance Monad (State s) where
...
 -- prop> \xs -> case firstRepeat xs of Empty -> True; Full x -> let (l, (rx :. rs)) = span (/= x) xs in let (l2, r2) = span (/= x) rs in let l3 = hlist (l ++ (rx :. Nil) ++ l2) in nub l3 == l3
-firstRepeat ::
-  Ord a =>
-  List a
-  -> Optional a
-firstRepeat =
-  error "todo: Course.State#firstRepeat"
+--
+-- attempt 1: my misconception of what was being asked because I couldn't
+-- read the test cases
```
Comment (JulianLeviston) [inline src/Course/State.hs:158]: Yeah, the instructions are pretty annoyingly unclear sometimes aren't they? It took me about 20 minutes to try to figure out exactly what the last one meant.

File: src/Course/State.hs [hunk 7]
```diff
@@ -135,44 +122,137 @@ instance Monad (State s) where
...
+--     firstRepeat1 (x :. rest) set
+--       | S.member x set = Full x
+--       | otherwise = firstRepeat1 rest (S.insert x set)
+
+-- attempt 3: using findM and a State
+--
+firstRepeat :: Ord a => List a -> Optional a
+firstRepeat xs = eval (findM p xs) S.empty
+  where
+    p x = State $ \set -> (S.member x set, S.insert x set)
```
Comment (JulianLeviston) [inline src/Course/State.hs:183]: Oh wow... this is delightfully elegant and clever! Mine is much longer, and doesn't make use of the fact that `S.insert x set == (S.insert x (S.insert x set))` (ie the fact that sets manage their own uniqueness).

Comment (JulianLeviston) [reply src/Course/State.hs:183]: An interesting alternate view, maybe... this is kind of the feeling of programming with `State` as a monadic interface...

```
firstRepeat xs = eval (findM mp xs) S.empty
  where
    mp x =
      get >>= \existingValues -> 
      put (S.insert x existingValues) >>= \_ -> 
      pure $ S.member x existingValues
```

It kind of reads off line an imperative program then...

File: src/Course/State.hs [hunk 8]
```diff
@@ -135,44 +122,137 @@ instance Monad (State s) where
...
+memberAndAddN x | trace ("memberAndAddN x = " P.++ show x) False = undefined
+memberAndAddN x = State (\set -> memberAndAddSetN x set)
+
+memberAndAddSetN :: Int -> S.Set Int -> (Bool, S.Set Int)
+memberAndAddSetN x set | trace ("memberAndAddSetN x = " P.++ show x P.++ ", set = " P.++ show set) False = undefined
+memberAndAddSetN x set = (S.member x set, S.insert x set)
+
+--
+-- at ghci
+-- let p x = get >>= \s -> put (Data.Set.insert x s) >>= \_ -> return (Data.Set.member x s) in runState (findM p (listh [1,2,3,1,2,3])) Data.Set.empty
```
Comment (JulianLeviston) [inline src/Course/State.hs:218]: ahhh cool... :)

File: src/Course/State.hs [hunk 9]
```diff
@@ -135,44 +122,137 @@ instance Monad (State s) where
...
+--
+-- simpler without bind:
+--
+-- let p x = State (\set -> (Data.Set.member x set, Data.Set.insert x set)) in runState (findM p (listh [1,2,3,1,2,3])) Data.Set.empty
```
Comment (JulianLeviston) [inline src/Course/State.hs:222]: this is really elegant :) I think previous greg might have found the monadic version easier to read (and maybe current Julian does, too)... but it's good either way.

File: src/Course/State.hs [hunk 10]
```diff
@@ -135,44 +122,137 @@ instance Monad (State s) where
...
+--     distinct1 Nil _ = Nil
+--     distinct1 (x :. rest) set
+--       | S.member x set = distinct1 rest (S.insert x set)
+--       | otherwise = x :. (distinct1 rest (S.insert x set))
+
+-- attempt 2: using filtering and State
+distinct :: Ord a => List a -> List a
+distinct xs = eval (filtering p xs) S.empty
+  where
+    p x = State $ \set -> (S.member x set, S.insert x set)
```
Comment (JulianLeviston) [inline src/Course/State.hs:250]: again, super awesome :)

File: src/Course/State.hs [hunk 11]
```diff
@@ -191,8 +271,39 @@ distinct =
...
+-- attempt 2: using firstRepeat, produce, join, contains
+isHappy :: Integer -> Bool
+isHappy n = contains 1 $ firstRepeat (produce sumOfSquareOfDigits n)
+  where
+    sumOfSquareOfDigits :: Integer -> Integer
+    sumOfSquareOfDigits m = sumIntegers (map square (digits m))
+    sumIntegers :: List Integer -> Integer
+    sumIntegers ms = foldRight (+) 0 ms
+    square :: Integer -> Integer
+    square = join (*)
```
Comment (JulianLeviston) [inline src/Course/State.hs:305]: wasn't that one a pain? guh... took me *ages* to work out `join (*)` was the `((->) r)` instance applied to itself. lol. What a head-bang.

File: src/Course/State.hs [hunk 12]
```diff
@@ -191,8 +271,39 @@ distinct =
...
+    digits :: Integer -> List Integer
+    digits m = map charToInteger (listh (show m))
```
Comment (JulianLeviston) [inline src/Course/State.hs:307]: nicely done. This was not what I would call easy...

File: src/Course/State.hs [hunk 13]
```diff
@@ -191,8 +271,39 @@ distinct =
...
+    charToInteger :: Char -> Integer
+    charToInteger = toInteger . Data.Char.digitToInt
```
Comment (JulianLeviston) [inline src/Course/State.hs:309]: very very cool 👍 especially considering they haven't really taken us through the numeric stack yet much...

File: src/Course/State.hs (hunk 2 above)
Comment (puyo) [reply src/Course/State.hs:66]: My fmap and apply in this code are a mess that somehow pass but are probably the least efficient possible.

I think part of me wonders whether ghc could/would optimise something like `g (f x) (f x)`, or in this case, `(f $ runState state x, g $ runState state x)`.

File: src/Course/State.hs (hunk 3 above)
Comment (puyo) [reply src/Course/State.hs:86]: I definitely like the `let` and the destructuring instead of `fst`, `eval`, or `exec`.

**Pull request merged**